### PR TITLE
feat: add topic parsing with # wild card

### LINKED ISSUE #10476: MQTT topic parsing with # wild card
## Feature Request

Opening a feature request kicks off a discussion.

### Use case:  
Has been asked for in the community slack https://influxcommunity.slack.com/archives/CH99HUH8V/p1642071313079200?thread_ts=1641120682.410600&cid=CH99HUH8V

### PR DESCRIPTION
**Pull request by MyaLongmire**
### Required for all PRs:

- [x] Updated associated README.md.
- [x] Wrote appropriate unit tests.
- [x] Pull request title or commits are in [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/#summary)

resolves #10476  [community issue](https://influxcommunity.slack.com/archives/CH99HUH8V/p1641323184432000?thread_ts=1641120682.410600&cid=CH99HUH8V)

Added topic parsing with # wild card by checking if it is used and then grabbing the length of that topic instead of the given parsed topic. This lessens the rule of topic and measurement, tags, fields, having to be the same length. Now topic can be equal to or longer in the number of elements.

### CODE REVIEW

File: plugins/inputs/mqtt_consumer/mqtt_consumer.go [hunk 1]
```diff
@@ -289,11 +291,16 @@ func (m *MQTTConsumer) recvMessage(_ mqtt.Client, msg mqtt.Message) {
 // compareTopics is used to support the mqtt wild card `+` which allows for one topic of any value
 func compareTopics(expected []string, incoming []string) bool {
 	if len(expected) != len(incoming) {
+		for i := range expected {
+			if strings.Contains(expected[i], "#") {
+				return true
+			}
+		}
 		return false
 	}
```
Comment (srebhan) [inline plugins/inputs/mqtt_consumer/mqtt_consumer.go:300]: How about
```suggestion
	if len(expected) > len(incoming) {
		return false
	}
	if len(expected) < len(incoming) {
		return choice.Contains("#", expected)
	}
```

File: plugins/inputs/mqtt_consumer/mqtt_consumer.go [hunk 2]
```diff
@@ -313,20 +320,32 @@ func (m *MQTTConsumer) onMessage(acc telegraf.TrackingAccumulator, msg mqtt.Mess
 		}
 		for _, p := range m.TopicParsing {
 			values := strings.Split(msg.Topic(), "/")
+			if len(values) < len(p.SplitTopic) {
+				return fmt.Errorf("config error topic parsing: found topic is shorter than parse topic length")
+			}
```
Comment (srebhan) [inline plugins/inputs/mqtt_consumer/mqtt_consumer.go:325]: Are you sure this is an error?
Think about a case where you have topic tree with e.g. `/location/device/value` at the one branch but another subtree `/location/device/status/{battery, network, ...}`  on the other branch. If you now apply topic parsing to the later (e.g. to `/location/device/status/+`), the `p.SplitTopic` will always be shorter for the `/location/device/value` topic. Am I missing something?

**Review by srebhan — COMMENTED** [review]
Very nice implementation @MyaLongmire! I just have one simplification (hopefully) and a question about handling topic branches of different length...

File: plugins/inputs/mqtt_consumer/mqtt_consumer.go (hunk 1 above)
Comment (MyaLongmire) [reply plugins/inputs/mqtt_consumer/mqtt_consumer.go]: What is `choice` in `choice.Contains`?

File: plugins/inputs/mqtt_consumer/mqtt_consumer.go (hunk 2 above)
Comment (MyaLongmire) [reply plugins/inputs/mqtt_consumer/mqtt_consumer.go:325]: I think it is an error as it is checking that the topic it found is shorter than the topic given. Say the config has the topic `telegraf/testing/#` but the topic that is is found is just `telegraf/testing`. What is found is shorter than what is given so it should error. I added a test for this case. 

Maybe I am misunderstanding your example. Thoughts?

Comment (srebhan) [reply plugins/inputs/mqtt_consumer/mqtt_consumer.go:325]: Hmm but in your example, what if there is another topic `telegraf/status`? It's also shorter and maybe you just don't want to parse it? Am I missing some previous check or something?

**Comment by MyaLongmire** [comment]
@powersj I think there are some fundamental problems with this approach. So I am going to close it for now. If someone wants to take it over that's cool with me.

File: plugins/inputs/mqtt_consumer/mqtt_consumer.go (hunk 2 above)
Comment (MyaLongmire) [reply plugins/inputs/mqtt_consumer/mqtt_consumer.go:325]: You have to explicitly state which topic you want parsed. So another topic that is shorter should not even get looked at.

**Pull request closed without merging**